repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java

```diff
@@ -122,6 +167,66 @@ private void processUpdateProduct(AtlasEntity entity, AtlasVertex vertex) throws
+
+            switch (currentProductDaapVisibility) {
+                case "Private":
+                    switch (newProductDaapVisibility) {
+                        case "Protected":
+                            // create policy for policyUsers and policyGroups
+                            policy.setAttribute(ATTR_POLICY_USERS, entity.getAttribute("daapVisibilityUsers"));
+                            policy.setAttribute(ATTR_POLICY_GROUPS, entity.getAttribute("daapVisibilityGroups"));
+                            break;
+                        case "Public":
+                            // set empty user list
+                            policy.setAttribute(ATTR_POLICY_USERS, Arrays.asList());
+                            // set user groups oto public to allow access for all
+                            policy.setAttribute(ATTR_POLICY_GROUPS, Arrays.asList("Public"));
+                    }
+                    break;
+                case "Protected":
+                    switch (newProductDaapVisibility) {
+                        case "Private":
+                            // create policy for policyUsers and policyGroups
+                            policy.setAttribute(ATTR_POLICY_USERS,Arrays.asList());
+                            policy.setAttribute(ATTR_POLICY_GROUPS,Arrays.asList());
+                            break;
+                        case "Public":
+                            // set empty user list
+                            policy.setAttribute(ATTR_POLICY_USERS, Arrays.asList());
+                            // set user groups oto public to allow access for all
+                            policy.setAttribute(ATTR_POLICY_GROUPS, Arrays.asList("Public"));
+                    }
+                    break;
+                case "Public":
+                    switch (newProductDaapVisibility) {
+                        case "Private":
+                            // create policy for policyUsers and policyGroups
+                            policy.setAttribute(ATTR_POLICY_USERS,Arrays.asList());
+                            policy.setAttribute(ATTR_POLICY_GROUPS,Arrays.asList());
+                            break;
+                        case "Protected":
+                            // create policy for policyUsers and policyGroups
+                            policy.setAttribute(ATTR_POLICY_USERS, entity.getAttribute("daapVisibilityUsers"));
+                            policy.setAttribute(ATTR_POLICY_GROUPS, entity.getAttribute("daapVisibilityGroups"));
+                            break;
+                    }
+                    break;
+            }
+            AtlasEntity.AtlasEntitiesWithExtInfo policiesExtInfo = new AtlasEntity.AtlasEntitiesWithExtInfo();
+            policiesExtInfo.addEntity(policy);
+            EntityStream entityStream = new AtlasEntityStream(policiesExtInfo);
+            entityStore.createOrUpdate(entityStream, false); // adding new policy
+        }
```

Comment:
Even here we can make this more modular by having a different function for policy updation.